straitlets/ext/click.py

```diff
@@ -0,0 +1,101 @@
+from __future__ import absolute_import
+
+from operator import itemgetter
+
+import click
+
+from straitlets import MultipleTraitErrors
+from traitlets import TraitError
+
+
+def _indent(msg):
+    return '\n'.join('  ' + line for line in msg.splitlines())
+
+
+class _ConfigFile(click.File):
+    def __init__(self, config_type, encoding=None):
+        super(_ConfigFile, self).__init__(
+            mode='r',
+            encoding=None,
+            errors='strict',
+            lazy=False,
+            atomic=False,
+        )
+        self.config = config_type
+
+    def _reader(self, f):  # pragma: no cover
+        raise NotImplementedError('_reader')
+
+    @property
+    def name(self):  # pragma: no cover
+        raise NotImplementedError('name')
+
+    def convert(self, value, param, ctx):
+        f = super(_ConfigFile, self).convert(value, param, ctx)
+        try:
+            return self._reader(f)
+        except MultipleTraitErrors as e:
+            self.fail(
+                'Failed to validate the schema:\n\n' + '\n'.join(
+                    '%s:\n%s' % (key, _indent(str(err)))
+                    for key, err in sorted(e.errors.items(), key=itemgetter(0))
+                ),
+                param,
+                ctx,
+            )
+        except TraitError as e:
+            self.fail(
+                'Failed to validate the schema:\n%s' % _indent(str(e)),
+                param,
+                ctx,
+            )
+
+
+class JsonConfigFile(_ConfigFile):
+    """A click parameter type for reading a :class:`~straitlets.Serializable`
+    object out of json file.
+
+    Parameters
+    ----------
+    config_type : type[Serializable]
+        A subclass of :class:`~straitlets.Serializable`.
+
+    Notes
+    -----
+    Normal :class:`~straitlets.Serializable` are not eagerly validated, if you
+    want to check the schema of the file at read time you should use a
+    :class:`~straitlets.StrictSerializable`.
+    """
+    name = 'JSON-FILE'
+
+    def _reader(self, f):
+        return self.config.from_json(f.read())
+
+
+class YamlConfigFile(_ConfigFile):
+    """A click parameter type for reading a :class:`~straitlets.Serializable`
+    object out of base64 encoded json file. This is often used with stdin input
+    which is specified as ``-`` on the command line.
+
+    Parameters
+    ----------
+    config_type : type[Serializable]
+        A subclass of :class:`~straitlets.Serializable`.
+
+    Notes
+    -----
+    Normal :class:`~straitlets.Serializable` are not eagerly validated, if you
+    want to check the schema of the file at read time you should use a
+    :class:`~straitlets.StrictSerializable`.
+    """
+    name = 'YAML-FILE'
+
+    def _reader(self, f):
+        return self.config.from_yaml(f)
+
+
+class Base64ConfigFile(_ConfigFile):
```

Comment:
I probably wouldn't even bother with this one.  I can't think of a circumstance where I'd want a base64 config file.  That's like using the headers of syntatically-valid HTTP requests as a config format.